### ISSUE TITLE
Implement internal Editorial Report UI and algorithms

### ISSUE THREAD

**Issue by jmacgreg** [pkp/pkp-lib#4844]
From PKP 2019 sprint; see [this specification](https://docs.google.com/document/d/19IYU37-SQIZNKNfGnPet1sN-xcJB63VcpFp96oVxyMo/edit#heading=h.406q5inyilkt) for more information. Discussion should happen in this issue.

**Comment by jmacgreg** [comment]
Hi Jonas, this is looking great so far! Some comments/fix suggestions: 

- Change the top "Editorial Activity" title to "Current Workflow Overview". 
- Remove the date picker option from the top chart, as I don't think it has any effect there. 
- For OJS, remove the "Internal Review" option, and rename "External Review" to simply "Review". For OMP (if we port this function over) the phrasing can stay the same, as OMP has internal and external review, but OJS doesn't. 
- limit the behaviour of the date picker and filter options to the chart or table that they "belong" to; or somehow otherwise streamline how these are applied. This may be a bit more difficult, I'm not really sure, and it could probably use a review by @NateWr. Right now there are multiple different locations to choose date spans and apply the filters - basically, the tools are available for each chart and table on the page, but apply across the whole page (where applicable) and that looks confusing. So perhaps I'd suggest one of the following: 
  - have the controls in one location only, at the top of the page, to be applied universally; OR
  - have the controls at the top of each chart/table, but only the controls that apply there, and have them only apply to that one chart/table.

**Comment by NateWr** [comment]
@jonasraoni do you mind opening a PR for OJS/pkp-lib for your changes? That will make it easier for me to do a review. I'll try to take a closer look this week.

**Comment by jonasraoni** [comment]
Done @NateWr! As @jmacgreg asked to develop this feature over `/stable-3_1_2`, I've opened the PRs against it... Let me know if I should change the target to `/master`.

I'll add some comments.

**Comment by NateWr** [comment]
Looks great, @jonasraoni! I haven't yet done a code review but the functionality is :ok_hand:. I've created a few pull requests that address some of @jmacgreg's comments.

PRs:
https://github.com/jonasraoni/pkp-lib/pull/1
https://github.com/jonasraoni/ojs/pull/1
https://github.com/jonasraoni/ui-library/pull/1

Here's how it is laid out after these PRs:

![editorial-activity](https://user-images.githubusercontent.com/2306629/63423755-e3aa1a80-c404-11e9-8de4-619bbe5e658d.png)

Changes:
- Only one set of filters/date range buttons.
- The filters/date range have been brought below the active submissions panel so it is clearer that they only apply to the tables.
- Adjustments to `<h*>` elements to provide an accessible heading structure for the page.
- The text `3 active submissions` is now an `<h2>`. @jmacgreg you suggested changing the top-level title to "Current Workflow Overview". But now that we've moved the graphic bit to the top, do you think the text "3 active submissions" clears up any confusion that existed before? Or do we still need a title to describe that panel?

Personally, I still think there's a little bit of confusion around the date range only effecting one of the columns. But I don't have a solution and it doesn't strike me as a major issue.

**Comment by alexxxmendonca** [comment]
One important detail that perhaps might not be in the [specification document](https://docs.google.com/document/d/19IYU37-SQIZNKNfGnPet1sN-xcJB63VcpFp96oVxyMo/edit#heading=h.406q5inyilkt): what exactly is the date range taking into account? Submission date or simply date?

My concern is, if the date range is taking into account the Submission date, the data for decisions (accept, reject, etc) will be limited to papers that have been submitted within that date range.

**Comment by NateWr** [comment]
We had a discussion about that. I believe it was decided that it will take the date of submission. So if the Date Range is January, 2019, the Acceptance Rate would mean:

> Out of 10 submissions received in January, 2019, three of them were later accepted, including those accepted after January.

Maybe James or Jonas can clarify.

**Comment by alexxxmendonca** [comment]
In a scenario where an Editor needs to know how many papers were Accepted in a given period of time, regardless of when they were submitted, can they have that information using the options provided?

**Comment by NateWr** [comment]
Ok, I'm making assumptions now, but my understanding is that would be available in "Submissions Accepted" near the top. So "Submissions Accepted" means "number of submissions accepted in this date range".

The one I described before was "Acceptance Rate", closer to the bottom. Sorry for the confusion. I should just defer to James or Jonas who will know better.

**Comment by jonasraoni** [comment]
@jmacgreg I've spread the filter everywhere because as an user I prefer to avoid scrolling the page to update filters/options (so an even better option for **me** would be fixed/floating controls). Also, as an user, I'd rather have the filter applied to both `Trends` and `User Registration`, but in case you prefer to add a separated filter for each one, no problem. There's another doubt for you below =]

@NateWr Thanks, I'll take a look soon.

@alexxxmendonca Let's confirm with @jmacgreg! Initially everything was using the submission date, but then I re-read this phrase:

> So when they suggest a date range, we only look at submissions that were submitted, and have a decision, for that date span.

... And modified the filter to use the date of the last decision.

So, currently the calculations with days (e.g. average days to decide) are using the submission date, but the date filter that controls which submissions are included in the statistics is using the decision date... This way a submission that was created 100 years ago and approved/declined within the date range would be counted.

**Comment by alexxxmendonca** [comment]
> This way a submission that was created 100 years ago and approved/declined within the date range would be counted.

Would be counted where, exactly? In the "Submissions Accepted" and "Submissions Declined"?

So:

"Submission Received" - takes into account the submission date
"Submissions Accepted", "Submissions Declined" - takes into account the decision date

Is that so?

**Comment by jonasraoni** [comment]
@alexxxmendonca FYI I've modified the code to use the submission date. In case you think it's useful to have the decision date as an option, we can improve the filter.

**Comment by NateWr** [comment]
> FYI I've modified the code to use the submission date. In case you think it's useful to have the decision date as an option, we can improve the filter.

I think there's some confusion here about two different rows in the table:

- Submissions Accepted
- Acceptance Rate

I believe @alexxxmendonca was hoping that "Submissions Accepted" would work based on the acceptance date. So a Date Range of "Last 30 days" would display the number of submissions with an accept editor decision in the last 30 days, regardless of when they were submitted.

The "Acceptance Rate" row is more confusing. But I think it makes sense here to take into account the submission date. So a Date Range of "Jan 1, 2018 to Dec 31, 2018" would display the percentage of submissions that were submitted in 2018 and eventually had an accept editor decision.

The same would be the case for Submissions Rejected and the Rejection Rates.

@alexxxmendonca does this line up with what you were thinking?

@jmacgreg I know we went over this a lot before, but is this what you had in mind or have I gotten off track?

(As a side note, in the `master` branch we have a `Tooltip` component which we can use to describe each row in the table. That way we can make these things clear.)

**Comment by alexxxmendonca** [comment]
Hi @NateWr,

> So a Date Range of "Last 30 days" would display the number of submissions with an accept editor decision in the last 30 days, regardless of when they were submitted.

Yes, this looks good to me.

> The "Acceptance Rate" row is more confusing. But I think it makes sense here to take into account the submission date. So a Date Range of "Jan 1, 2018 to Dec 31, 2018" would display the percentage of submissions that were submitted in 2018 and eventually had an accept editor decision.

It is a bit confusing, since the acceptance can take up to 1 year or more after the submission date, depending when you run the report, you might get a very low acceptance rate. These figures would be more faithful when the date range is from at least 2 years prior and before -- maybe this should be in the tooltip too?

But the logic is correct, take into account the submission date.

It's important to not consider the same portion of submitted papers to make the accept

> (As a side note, in the master branch we have a Tooltip component which we can use to describe each row in the table. That way we can make these things clear.)

I was going to suggest something along these lines. We know now what it all means but we have to leave that clear enough for users.

**Comment by jonasraoni** [comment]
@NateWr I understood what @alexxxmendonca wanted, but I thought it was better to revert and keep what I (mis)understood from that email, until there's a confirmation from @jmacgreg (I don't know who's the stakeholder 😁).

**Comment by jonasraoni** [comment]
Hi guys, FYI I'm back to the review! I've already modified the queries locally to "detach the dates", but I'll update some internal things before coming up with news.

**Comment by NateWr** [comment]
We're almost there! New PRs for the stable branch. Tests running now.

PRs:
https://github.com/pkp/pkp-lib/pull/5274
https://github.com/pkp/ui-library/pull/51
https://github.com/pkp/ojs/pull/2530

**Comment by NateWr** [comment]
@asmecher it occurred to me that this feature requires the upgrade script run, and I believe you've said you don't want upgrade things going into stable releases:

https://github.com/pkp/ojs/pull/2530/files#diff-02fab074e950f9db1ece033a61599d30R232

What do you want to do about this? I think this is only for the scheduled task to send an email with the journal stats to a manager. So perhaps there will be no problem if the upgrade script isn't run, but I'm not sure.

**Comment by NateWr** [comment]
I've done some checking to see if the stats returned are lining up with expectations and they're pretty good! Where we have some real problems is in the Average/Year column, mostly down to confusion about how we calculate a year.

## Needs a cut-off when calculated averages

The following rows return data based on the date of the activity reported. So "Submissions Accepted" returns the count of submissions that had an accept decision within the requested date range:

- Submissions Received
- Submissions Accepted
- Submissions Declined
- Submissions Declined (Desk Reject)
- Submissions Declined (Post-Review)
- Submissions Declined (Other)
- Submissions Published

These rows return an Average/Year. However, the way the average is calculated can be misleading. The total is divided by the number of years since the first submission: (120 submissions / 4 years). This means that the average will adjust based on what time of year we are making the request. Consider the following:

> A journal has received 120 submissions since 2016. On Dec 31, 2019, the average per year is 30 (120 submissions / 4 years). No new submissions occur before I check again on Jan 1, 2020, but the average per year is now 24 (120 submissions / 5 years).

I think we need to set a cut-off point and only calculate the average from that. For example, we can say that only submissions for completed years are included in the average calculation, and leave the current year out. So in 2019 it would be (All submissions before Jan 1 2019 / Number of years from first submission to 2018).

We have a similar problem with the start date. Consider this case:

> A new journal began taking submissions in Nov 2017. They have 100 submissions. When checking checking their average in January, 2019 they are shown an average of (100 submissions in 2017 and 2018 / 2 years). The average says 50, but that's not accurate because only one year received submissions for the whole year.

I can think of a few ways to handle this:

1. Leave out the first year to skip incomplete years.
2. Calculate when the first submission was made and adjust the total for that year to compensate for the "missed days".
3. A combination of the two. We leave out the first year unless we find that the first submissions occurred in the first month or two of that year, in which case we include it.

I'm tempted to lean towards 1.

## Average/Year doesn't make sense

The following rows include an Average/Year, but I think that this column doesn't make sense for them. Because they are already averages, the "Total" column is the only one needed.

- Average Days to First Editorial Decision
- Average # Days to Accept
- Average # Days to Reject
- Acceptance Rate
- Rejection Rate
- Desk Reject Rate
- Post-Review Reject Rate
- Other Reject Rate

The Average/Year is always the same as the Total. I think these rows should just have a `-` in the Average/Year column, or it should be left empty.

## Users

The Average/Year column is always `0` in the users table. All of the same things apply to how it would be calculated if it were fixed.

I also think maybe there's a bit of confusion about what the user registrations mean during a date range. It isn't a count of the number of journal managers during a date, but of how many signups there are.

Maybe we should change the heading to something like "New User Registrations" and change the "Total Within Date Range" to "New Signups Within Date Range"?

**Comment by NateWr** [comment]
I forgot to say that I think to get this into the next stable release, we should consider removing the Average/Year column. We can restore it in 3.2 once we have figured out these issues.

**Comment by jmacgreg** [comment]
Hi Nate - agreed with you here, and I'm comfortable with leaving the Average/Year column out for now.

Your suggestions for the user registrations bit make sense to me as well.

**Comment by alexxxmendonca** [comment]
> I forgot to say that I think to get this into the next stable release, we should consider removing the Average/Year column. We can restore it in 3.2 once we have figured out these issues.

I would agree with this approach. In fact, reading all your thoughts on average it's making me rethink whether we should offer them. We might just leave that for people to calculate it using their favorite spreadsheet software.

If we do go towards that path, it is important to clearly describe the rule adopted to calculate these averages (using a tooltip or something).

Is the average calculated depending on the date range? If so, personally I wouldn't change anything about the general average rule, but perhaps would include a clear instruction saying that for averages, results will look better when working with full years, or something along these lines.

**Comment by NateWr** [comment]
> Is the average calculated depending on the date range?

No, not at the moment. Instead, the average is always calculated from the total divided by the number of years since the first submission.

I will adjust the stable PRs to remove averages and file a separate issue to work on averages in the future. @jmacgreg do the averages need to be part of 3.2 or can this issue be filed for later work. I'm thinking particularly about the timeline for feature-freeze for 3.2.

**Comment by jmacgreg** [comment]
Hi @NateWr, the averages can be filed for later work - no need to force this stuff to be included in 3.2 given that it definitely needs further UX review.

**Comment by NateWr** [comment]
Filed here: https://github.com/pkp/pkp-lib/issues/5275

**Comment by NateWr** [comment]
Merged to stable without averages. Working on the port to `master` now.

**Comment by asmecher** [comment]
@NateWr, on [upgrade scripts in builds](https://github.com/pkp/pkp-lib/issues/4844#issuecomment-553952771), that's correct, we want to avoid that. We chose that policy so that git checkouts across any version on a stable branch would be DB-compatible. While your example isn't a schema change per se, I suspect users will (and perhaps should) feel OK about doing a code update for a new build without running the upgrade script at all, and adding an upgrade stanza undermines that.

If an email template is missing, OJS will send out a blank template.

So even though this isn't a "breaking" (i.e. schema-altering) database change, I think I'd feel better if we didn't introduce it -- would it be possible to simply not use that email template in the 3.1.2 stable branch?

**Comment by NateWr** [comment]
@asmecher Looking more closely, this email template is invoked by the monthly editorial report email that is sent out automatically. This is triggered by a new scheduled task.

Am I right in thinking that if no database upgrade is run the scheduled task will never be set up and so the email template does not need to be run? In this case, we can leave it in for people who run the upgrade script but it will have no impact on people who do not.

That might work well for the PS clients who are eager for this. But if I'm mistaken then I can take it out of stable.

**Comment by asmecher** [comment]
You're right that if the upgrade script isn't run, the task won't be added. So technically it's OK but IMO the decision on whether to include this hinges on whether the "promise" to not script database upgrades between builds is worth breaking; I'm still leaning towards no, though I can be swayed. Part of my reasoning is that we're eventually going to move from our current upgrade toolset to migrations, and this case would be an outlier.

**Comment by asmecher** [comment]
I worked around the problem with OJS 3.1.2-2 by removing the entry from `registry/scheduledTasks.xml` (and removing the database update clause from `dbscripts/xml/upgrade.xml`), but leaving the task class in place. This does mean that users who ran the upgrade script to 3.1.2-2 will need to intervene manually to remove the entry from `scheduled_tasks` (or modify the task class per https://forum.pkp.sfu.ca/t/ojs-3-1-2-2-editorial-report-emails/57243/3), even if they move to 3.1.2-3, but since 3.1.2-2 was out for such a short time I think that'll be a very small impact.

For the release that activates the task, several things need attention:
- The upgrade script clause had the wrong version range
- Newly-added notifications should generally be opt-in, rather than assuming that upgrading users will want to receive them
- The high volumes of email that this apparently sent need investigation.

@NateWr, I had spoken about filing a separate issue for this, but in retrospect I think this one is enough.

**Comment by NateWr** [comment]
Progress is coming along with the merge to `master`. I have a few more questions about how we should be calculating the stats:

1. **Declined (Other)** - What is the use case here? Are there decision records from old versions that we are trying to accommodate? Without a specific decision to look for, there is no way to calculate the number of "Declined (Other)" submissions for a particular date range.

2. **Declined** - We have a similar problem when calculating the number of declined submissions in a date range. Without an actual editor decision record to look for, we don't have any record of when a submission was declined. We can calculate the total number of declined submissions but not those declined within a date range. We can only add Declined (Desk-Reject) and Declined (Post-Review).

3. **Acceptance Rate** and **Rejection Rate** - We've discussed how these figures can be misleading because not all submissions made during a given period have yet had a chance to be accepted or rejected. That's because we do the following to calculate the rate: `rate = count of accepted / total received`. The `total received` number will include submissions that have been neither accepted or rejected. Should we instead calculate the `total received` so that it only includes submissions that have been accepted or rejected? I think that would make this number accurate regardless of what date range was applied -- though for short recent date ranges like the last month it will likely just be `0`.

Finally, this is really an 11th hour suggestion, but can we change the "Average Days" statistic to something like "80% of submissions reach first decision within X days"? Two reasons. First, an average can be thrown off easily by a few outliers. In the test data I'm looking at, it's clear some authors disappeared after review and so the reject decision happened over a year after submission, probably when an editor finally got around to clearing out stale submissions. 

Second, if we presume that editors will use this statistic to communicate to authors and stakeholders, an average can be problematic. For example, if I tell authors that 27 days is the average to first decision, by necessity about 50% of authors will be disappointed with their response time. They'll think "I should hear a response in 27 days" when actually the stat means they're just as likely to not hear a response in that many days.

Changing the average to a median will solve the first problem, but I really think that addressing the second problem will give us a more meaningful stat. As an editor, if I want to know what the author experience is like for my journal, it matters more what the majority of authors experience rather than what the average is. I think that a statistic like 80% of authors hear from us in 30 days is a much more honest stat that frames the common experience (rather than the average) as the priority.

**Comment by NateWr** [comment]
> I also think maybe there's a bit of confusion about what the user registrations mean during a date range. It isn't a count of the number of journal managers during a date, but of how many signups there are.

I want to revisit this, because I think it could be quite misleading. It's not much of a problem for the roles like readers and authors, where most of the time they'll be assigned the role when they register an account.

But when we get down to subeditors, assistants and managers, it gets pretty messy. For these users, there is very likely going to be some historical evolution of their roles in the journal. For example, a scholar may author a piece in 2015, be asked to review in 2016, and be asked to be a subeditor in 2018. In this case the stats would consider this user a "new subeditor" in 2015.

This is especially problematic because for these roles the numbers are often very low, so the counts of user signups for each year will be especially distorted.

For this same reason, the "Total" column is not really a total of signups. It actually identifies how many users have that role in the system right now. 

We don't, to my knowledge, have any record of when a user was assigned a role, so I'm not sure we can provide any historic data here. Even the general date that they signed up could be wrong in a multi-journal context where users are registering for one journal before registering for another.

**Comment by jonasraoni** [comment]
Hi @asmecher, about the third item: On Sunday I saw that the scheduler [has a check to run tasks that were missed](https://github.com/pkp/pkp-lib/blob/master/classes/scheduledTask/ScheduledTaskHelper.inc.php#L234), but it doesn't have a check to avoid running them again. So if you have a task scheduled for the "day 1", it might be executed N times (whenever the cron triggers the runScheduledTasks) until the day is over.

I've been using "cron parsers" when I need such feature: https://github.com/mtdowling/cron-expression (I've never used this one, it's just the first repository that appeared).

**Comment by alexxxmendonca** [comment]
> can we change the "Average Days" statistic to something like "80% of submissions reach first decision within X days"?

I see/understand your reasons and I'm okay with that. They would have access to the raw data in case they want to make the avarage they want.

**Comment by asmecher** [comment]
> Hi @asmecher, about the third item: On Sunday I saw that the scheduler has a check to run tasks that were missed, but it doesn't have a check to avoid running them again. So if you have a task scheduled for the "day 1", it might be executed N times (whenever the cron triggers the runScheduledTasks) until the day is over.
>
> I've been using "cron parsers" when I needed such feature: https://github.com/mtdowling/cron-expression  (I've never used this one, it's just the first repository that appeared).

There are two mechanisms that execute the scheduled tasks: the Acron plugin, and the scheduled tasks tool (via cron or equivalent). I recommend cron, as it's more reliable. The Acron plugin piggybacks on a request to perform the tasks, but it should only be considered a work-around for when cron isn't available.

When the scheduled tasks complete, the "date last run" is updated, so trying to run them again won't cause redundant messages. However, if the cron job fires while the last run is still in progress, it'll run the same tasks again. Setting cron to run daily (as recommended in the docs) should be enough to avoid this. Something like a fatal error during the tasks run will also prevent the "date last run" from being updated, thus the next cron run will potentially duplicate the part of the run that did succeed before the error.

As for the acron plugin, there is a potential race condition there -- if several requests come in quick succession, each may try to be the "mule" request that runs the tasks. This is partially mitigated by updating the "date last run" early, but still depends on the site being relatively low volume to avoid the risk. (When this code was written, there really wasn't a reliable locking mechanism available to PHP. See https://pkp.sfu.ca/bugzilla/show_bug.cgi?id=8737 for details.)

It's definitely a bad idea to have both Acron and cron enabled.

**Comment by alexxxmendonca** [comment]
Here's another use case following the 3.1.2.3 release:

![image](https://user-images.githubusercontent.com/4059679/70345201-37b4ff80-183a-11ea-87b1-d2bc2757f086.png)

As an editor, I would like to be able to know the average days between submission and final decision (Accept or Decline). The way the table is currently configured you have two separate averages -- one can't simply sum them to have an overall average, that would give them an incorrect average value. So I think we need an overall submission to final decision (by "final" it means either accept or reject).

Also, the "Desk Reject" is probably being taking into account on the **Average # Days to Reject** line, but as an Editor, I would like to be able to discount those from the average count, given that they might affect the resulting value (Desk Rejects tend to decrease your average since they are rejected within only a few days max after they are submitted).

**Comment by alexxxmendonca** [comment]
This is what a journal Editor has to go through if they wish to find out the average between submission and final decision: https://youtu.be/-TmYkYwXv3U (the video is in portuguese but one can have an ideia of the workload).

**Comment by NateWr** [comment]
Thanks @alexxxmendonca! We're already way behind schedule on getting this as-is into the `master` branch for v3.2, so I've filed that as a separate issue. But it's something we could add in 3.2.1 for sure.

**Comment by alexxxmendonca** [comment]
Excellent, @NateWr !

**Comment by NateWr** [comment]
PRs for the `master` branch:
https://github.com/pkp/ojs/pull/2546
https://github.com/pkp/omp/pull/732
https://github.com/pkp/pkp-lib/pull/5342
https://github.com/pkp/ui-library/pull/53

These PRs include a number of changes in line with the discussion above, as well as a few other judgement calls that I made. This means that stats will change when upgrading from 3.1.2-4 to 3.2. Here's a list of the changes to bear in mind:

- All managers and subeditors are opted out of the monthly email during the upgrade process. If they want to receive it, they will have to opt in by going to their Profile > Notifications and un-checking the box that disables it.
- Editorial Trends and Users are now shown on separate pages.
- Filtering user counts by date range has been removed (see [comment](https://github.com/pkp/pkp-lib/issues/4844#issuecomment-561294307)).
  - However, the API accepts `registeredAfter` and `registeredBefore` params so this information can be retrieved if desired.
- The "Declined (Other)" stat has been removed (see [comment](https://github.com/pkp/pkp-lib/issues/4844#issuecomment-561147202) and the "Declined" stat calculation has been changed to reflect that it just adds desk reject and after review stats.
- Published submissions within a date range only count the initial publication date -- not subsequent versions.
- Average days to decision have been changed to "Days to decision" and use an 80% threshold, so they represent: 80% of submissions that receive X decision do so within this many days. (See [comment](https://github.com/pkp/pkp-lib/issues/4844#issuecomment-561147202))
  - However, the average days can still be easily calculated with a method in the editorial stats service class.
- The acceptance and decline rates only count submissions that have received an accept/decline decision, so they exclude submissions still in the queue.
- I modified the [default email template](https://github.com/pkp/ojs/pull/2546/files#diff-ede2aeb2dfb88eecd563b9c5fef5209b) language.
- The available preset date ranges have been changed to reflect the longer timeline desired for these stats. They are:
  - Last 90 days (default)
  - Year to date
  - Last year (ie - 2018)
  - Last two years (ie - 2017/2018)

Also, I have added tooltips to the stats that I felt required them:

![editorial-stats](https://user-images.githubusercontent.com/2306629/70618827-aa8ff300-1c0b-11ea-9dc7-de2a1463dd33.png)
![editorial-stats-tooltip](https://user-images.githubusercontent.com/2306629/70618830-ab288980-1c0b-11ea-85b7-a02ed6b4c54b.png)

Please check the tooltip descriptions to see if they are clear: https://github.com/pkp/pkp-lib/pull/5342/files?file-filters%5B%5D=.po#diff-57d6778cfb81e9d47ac9ae5d7b33af24R1873-R1882

Finally, I'm unable to test the attachment in the monthly email because I don't have a local email server set up. Once this is merged I'll create some Acceptance Criteria for testing, but the email in particular needs some careful testing.

**Comment by NateWr** [comment]
This work is now merged to the `master` branch.
